src/essreflectometry/orso.py

```diff
@@ -1,15 +1,173 @@
-import warnings
+"""ORSO utilities for reflectometry."""
+import os
+import platform
+from datetime import datetime, timezone
+from typing import NewType, Optional
 
+import sciline
+from dateutil.parser import parse as parse_datetime
+from orsopy.fileio import base as orso_base
+from orsopy.fileio import data_source, orso, reduction
 
-def not_found_warning():
+from .types import Filename, RawData, Reference, Run, Sample
+
+
+class OrsoExperiment(
+    sciline.Scope[Run, data_source.Experiment], data_source.Experiment
+):
+    """ORSO experiment for a run."""
+
+
+class OrsoInstrument(
+    sciline.Scope[Run, data_source.InstrumentSettings], data_source.InstrumentSettings
+):
+    """ORSO instrument settings for a run."""
+
+
+class OrsoOwner(sciline.Scope[Run, orso_base.Person], orso_base.Person):
+    """ORSO owner of a file."""
+
+
+class OrsoReduction(sciline.Scope[Run, reduction.Reduction], reduction.Reduction):
+    """ORSO measurement for a run."""
+
+
+class OrsoSample(sciline.Scope[Run, data_source.Sample], data_source.Sample):
+    """ORSO sample of a run."""
+
+
+OrsoCreator = NewType('OrsoCreator', orso_base.Person)
+"""ORSO creator, that is, the person who processed the data."""
+
+OrsoIofQDataset = NewType('OrsoIofQDataset', orso.OrsoDataset)
+"""ORSO dataset for reduced I-of-Q data."""
+
+OrsoDataSource = NewType('OrsoDataSource', data_source.DataSource)
+"""ORSO data source."""
+
+OrsoMeasurement = NewType('OrsoMeasurement', data_source.Measurement)
+"""ORSO measurement."""
+
+
+def parse_orso_experiment(raw_data: RawData[Run]) -> OrsoExperiment[Run]:
+    """Parse ORSO experiment metadata from raw NeXus data."""
+    return OrsoExperiment(
+        data_source.Experiment(
+            title=raw_data['title'],
+            instrument=raw_data['name'],
+            facility=raw_data['facility'],
+            start_date=parse_datetime(raw_data['start_time']),
+            probe='neutron',
+        )
+    )
+
+
+def parse_orso_owner(raw_data: RawData[Run]) -> OrsoOwner[Run]:
+    """Parse ORSO owner metadata from raw NeXus data."""
+    return OrsoOwner(
+        orso_base.Person(
+            name=raw_data['user']['name'],
+            contact=raw_data['user']['email'],
+            affiliation=None,
+        )
+    )
+
+
+def parse_orso_sample(raw_data: RawData[Run]) -> OrsoSample[Run]:
+    """Parse ORSO sample metadata from raw NeXus data."""
+    if not raw_data.get('sample'):
+        return OrsoSample(data_source.Sample.empty())
+    raise NotImplementedError('NeXus sample parsing is not implemented')
+
+
+def build_orso_measurement(
+    sample_filename: Filename[Sample],
+    reference_filename: Optional[Filename[Reference]],
+    instrument: Optional[OrsoInstrument[Sample]],
+) -> OrsoMeasurement:
+    """Assemble ORSO measurement metadata."""
+    # TODO populate timestamp
+    #      doesn't work with a local file because we need the timestamp of the original,
+    #      SciCat can provide that
```

Comment:
Can't we use the `NXentry/start_time`?

Reply:
No. Because the timestamp is the last modification time. I.e., earliest the end time of the experiment, not the start time. And possibly later if the file was modified for some reason.

Reply:
Then use the end time? Modifying Raw files is not allowed, to my knowledge.

Reply:
There is no end time in our Amor file. I could add something now that gets the time if it exists and falls back to `None` otherwise. But I would rather wait for a more general mechanism in scippneutron that could also handle scicat timestamps and select the most appropriate one.